### ISSUE TITLE
Implement 0.4 tiles in existing components.

### ISSUE THREAD

**Issue by sandstrand** [Legofarmen/tunics#40]
- Figure out how we can have mulitlayered rooms.

The easiest way should be to allow the different components to place their own floor.
This requires a rethinking of how floors are output and will require a new approach to primary and secondary tiles.

**Comment by sandstrand** [comment]
#95 will allow for layered room. 

This issue should deal with placing of the correct floor placeholders in all components where applicable and updating all other tile id's in scripts to the new standard.

**Comment by sandstrand** [comment]
Renaming all the existing tiles in all components will be no easy task. There are 671 tiles in the old tileset.

If we want to be able to do this by sed, could you supply me with a proper syntax to use in the diff file I'm compiling. Something like
`"oldTileID" > "newTileID",`

There will also be some tiles were we want to evaluate if they are used properly or if the should be replaced by placeholders. I will exclude these from this file and compile another list.

Some tiles do not exist in the new tileset and should not be output att all. How should I name these?

**Comment by mattias-p** [comment]
To rename all occurrances of a tile id:

```
sed -i 's/"oldTileID"/"newTileID"/' file.dat
```

To remove all lines with a given tile id:

```
sed -i '/"unwantedTileID"/d' file.dat
```
- `-i` means modify file in-place
- `s///` means substitute
- `//d` means delete

If you remove the quotation marks, you can match partial ids.

**Comment by sandstrand** [comment]
I've created the following gists for batched replacement of tailes in existing components:

**diff_delete**
https://gist.github.com/sandstrand/130c650084f35197e5b1878a9b69f068
All occurrences of tiles with these ids should be removed, as `oldID`

**diff_replace**
https://gist.github.com/sandstrand/1d67933c0552de6a50b5635032b5d8a9 
All simple replacements, as `oldID, newID`

**diff_check**
https://gist.github.com/sandstrand/5e7753a8605ae108d2c2480fe742565b
Special cases where replacement is conditional upon usage and when composite placeholders should replace tiles. as `oldID` or `oldID, newID` depending on case.

They should be executed in the order listed here. I have not taken in to account that the script files for many components need to be updated as well, maybe it as simple as running the diff_replace against these as well.

I am not sure that all oldID exist in current components and there may some spelling errors with newIDs.

**Comment by mattias-p** [comment]
The grep-script wasn't very useful to me, and I ended up mashing everything else up into this. I call my script `runsed.sh`.

```
#!/bin/sh
compile () {
    sed -E -e '
    s:\s*$::
    s:^//:#:
    s:^([a-z0-9._]+),\s([a-z0-9._]+)$:s/(pattern|tileset) = "\1"/\\1 = "\2"/:
    s:^([a-z0-9._]+)$:/"\1"/d:
    '   
}
SCRIPT="
:loop
/}/b done
N
s/\n//
b loop
/^$/d
:done
`compile < diff_delete`
`compile < diff_replace`
s/,/,\n/g
s/\{/{\n/g
s/\}/}\n/g
/^$/D
"
find "$@" -name '*.dat' -exec echo {} \; -exec sed -i -E "$SCRIPT" {} \;
```

It reads the `diff_delete` and `diff_replace` from the current directory. The only things I've checked in are changes that aren't automated by this script. I made a few changes to `diff_replace` and a few to the `dungeon.tower_of_hera.dat` - hopefully correct ones.

Edit 1: Added pre- and post processing steps to remove entire entities and not just patternid attributes from entities.
Edit 2: Replaced lots of small scripts with a big one.

**Comment by mattias-p** [comment]
Do you have a sensible environment to do this in, by the way? Otherwise I can make a pull request into your pull request.

**Comment by mattias-p** [comment]
Apparently it's a bit more complicated. Shame on me for not checking.

**Comment by mattias-p** [comment]
I've got pretty far with the tileset switch. Some the above checks remain to be done, placeholders need to be taken better care of. Now I need to sleep. I'll try to get back to you tomorrow.

**Comment by sandstrand** [comment]
I can't really say I have a good environment for this. You might as well check in the transformed components into my pull request: https://github.com/sandstrand/tunics/pull/2

I also realised we may want to sed all current tilesets in components, and rooms, to tower_of_hera, in case you havent thought of that.

**Comment by mattias-p** [comment]
I pushed the transformed components, but I don't have write access to you repo, so you'll have to push the merge button.

**Comment by sandstrand** [comment]
Done. Checked and adjusted against diff_check for the following:
- [x] wall_hole
- [x] wall_statue.6
- [x] water_shallow
- [x] floor.20
- [ ] entrance_
- [x] hole
- [ ] bow_target

Not sure what you intend to do with the component placeholders so I leave them to you. The good news is that resizing tiles in the tileset does not need a review. Holes look just fine.

We now have a merge conflict in the pull request, not sure why, there was no conflict when merging your pull request into my 0.4_gfx branch. Did you rebase before creating the pull request?

**Comment by sandstrand** [comment]
This is messy. I'm considering pushing a properly rebased branch up until 'Transform components to use the new tileset structure (723e958f9b22f1300ea835e4b7103529bf65926e), which is where the conflict emerges, to Legofarmen/develop, along with the diffs. 

I've made some modifications to the diffs as well. We can redo the transformation there and I might be able to cherry pick my later commits and tweaks.

Or du you feel a new branch on Legofarmen would be better, if we want to keep develop semi-stable?

**Comment by mattias-p** [comment]
Yeah, throw out the transform commit and rebase. If you like I can make a new transform commit after that. I don't mind working against your branch, but you'll have to keep it rebased.

Is there an up to date `mappings.lua` available somewhere? I'd believe I need that to go forward with the placeholders.

**Comment by sandstrand** [comment]
RIght, I have a new branch: https://github.com/sandstrand/tunics/tree/0.4 rebased to 0.3 and with the diffs and all your changes up until transformation, i think. The latest mappings.lua in this branch only uses tower_of_hera but has no other 0.4-mappings.

The gist is as far as we've gotten: https://gist.github.com/sandstrand/5491b134a44789296bf0382d64afec24

Go ahead and transform again.

**Comment by sandstrand** [comment]
Pushed: all components checked and adjusted for diff_check. I think the only thing making it non-playable are the tiles which should be replaced by component placeholders.
